### PR TITLE
Collapse folders with only one child folder

### PR DESCRIPTION
**Pull request by matthiasthieroff**
## Proposed changes

Collapses a folder in code view which only has another folder as its only child. This lets you access a sub-folder which has content directly instead of navigating down the folder tree by clicking every folder separately.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [x] Related issues linked to PR if existing and labels set
- [x] New code is covered with unit tests
- [ ] New ui components are tested inside the storybook (module ui-components only) 
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog`
- [ ] Feature has been tested with different permissions
- [ ] Documentation updated (only necessary for new features or changed behaviour)

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-webapp/src/test/java/sonia/scm/api/v2/resources/SourceRootResourceTest.java [hunk 1]
```diff
@@ -24,7 +24,6 @@
 
 package sonia.scm.api.v2.resources;
 
-import com.google.inject.util.Providers;
```
Comment (pfeuffer) [inline scm-webapp/src/test/java/sonia/scm/api/v2/resources/SourceRootResourceTest.java:27]: I like the overall cleanup of this class, but somewhere you could add a further test method that checks something like

```java
verify(browseCommandBuilder).setCollapse(true)
```

**Review by pfeuffer — CHANGES_REQUESTED** [review]
Somehow this does not work with HG in some constellations. I have an old copy of the hg repository of SCM-Manager (get it [here](https://next-scm.cloudogu.com/scm/repo/pfeuffer/scm-manager-hg/info)).

Otherwise there are only minor remarks.

File: scm-core/src/test/java/sonia/scm/repository/api/BrowserResultCollapserTest.java [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    createFile(root, ".editorconfig");
+
+    BrowserResult result = new BrowserResult("revision", scmServer);
+    BrowseCommandRequest request = new BrowseCommandRequest();
+
+    new BrowserResultCollapser().collapseFolders(browseCommand, request, result.getFile());
+
+    FileObject f = result.getFile();
+    Collection<FileObject> children = f.getChildren();
+    assertThat(children.size()).isEqualTo(3);
```
[the remark this answers is not in the document]
Comment (matthiasthieroff) [reply scm-core/src/test/java/sonia/scm/repository/api/BrowserResultCollapserTest.java:263]: Unfortunately, I don't got a warning :|

File: scm-core/src/main/java/sonia/scm/repository/api/BrowseCommandBuilder.java [hunk 3]
```diff
@@ -152,6 +155,9 @@ public BrowserResult getBrowserResult() throws IOException {
         }
 
         result = browseCommand.getBrowserResult(request);
+        if (!request.isRecursive() && request.isCollapsed()) {
+          new BrowserResultCollapser().collapseFolders(browseCommand, request, result.getFile());
+        }
```
[the remark this answers is not in the document]
Comment (matthiasthieroff) [reply scm-core/src/main/java/sonia/scm/repository/api/BrowseCommandBuilder.java:160]: Do you have any suggestion for a good method name if you also extract the condition?

[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-core/src/main/java/sonia/scm/repository/api/BrowseCommandBuilder.java]: I thought of extracting the complete result computation, not only the `if` statement.

Something simple like

```java
result = computeBrowserResult(request);
```

[the remark this answers is not in the document]
Comment (matthiasthieroff) [reply scm-core/src/main/java/sonia/scm/repository/api/BrowseCommandBuilder.java]: Okay, commenting on the if statement was a bit misleading.

**Pull request merged**